### DISCUSSION TITLE
Image potential mistake - Block ciphers

### DISCUSSION THREAD

**Discussion by gonzalo-munillag** [OpenMined/courses#317] [Course: Foundations of Private Computation]
Hi :)

Amazing prep from Sebastià.

I think there might be a mistake in the images depicting the block cipher.
 I think **K** does not necessarily need to be lower than **N**, as a matter of fact, I think the last byte N should be in the last block. I think the number of blocks is round_down(**N**/**K**)+1. (With round down I mean that we forget about the reminder). In the image, however, the blocks go beyond N.

See this example from the notebook:
bytes per block: 16 = K
bits per block: 128
message length: 46 = N (This is in bytes)
padded_message lenght: 48 = 16***3**

We have 3 blocks

There are **N**/**K**+1 = round_down(46/16)+1=2+1=3

<img width="908" alt="Screenshot 2021-04-23 at 15 49 02" src="https://user-images.githubusercontent.com/57599753/115897275-89285380-a45c-11eb-89ea-1f32544d2f7e.png">

**Comment by hwrdtm** [comment]
Hi @gonzalo-munillag, not sure what you mean with this

>In the image, however, the blocks go beyond N.

I think the illustration is arbitrarily showing the ciphertext to have blocks of K bytes (in blue) as a generic depiction, and that it should not be taken literally that it is a longer byte array than the message itself?

You are correct though, given K = 16 bytes, a message have any length <K or >K.
